### DISCUSSION TITLE
Should plugins be configured inside rezconfig.py?

### DISCUSSION THREAD

**Discussion by bpabel** [AcademySoftwareFoundation/rez#1382] [New Features]
The recent PR that introduces a package repository plugin for S3 brought up a question for me regarding plugin configuration  -- https://github.com/AcademySoftwareFoundation/rez/pull/1380 

The PR in question requires the user to create a yaml configuration file called `settings.yaml` next to the plugin (which if the PR is accepted, would end up having to be inside the rez installation, along with the default set of plugins).

There's a somewhat undocumented feature of the rez config to provide configuration options for *plugin types*, but not necessarily for specific plugins. 

Should the rez config also be used to provide configuration options for specific plugins? And should that be the preferred method instead of having plugins use their own, separately managed configuration files?

My rationale:

Pros:

- Prevents differences in how configuration is done between default plugins and plugins maintained as a separate project, since options for default plugins are generally standardized into the plugin type and are configured via `rezconfig.py`
- Allows plugin options to be self-documenting along with the other `rezconfig.py` settings.
- There's already a mechanism to merge multiple `rezconfig.py` files, allowing plugins to distribute default configurations for themselves that can be added to the `REZ_CONFIG_FILE` path.

Cons:
- Forced or inflexible standardization on python as the config format, vs yaml, toml, or some other data format.

**Comment by maxnbk** [comment]
It's worth remembering when thinking about the direction here, that rez-config also has a way to defer_load fragments of config to other files. (Not sure that if that's what you meant by the merge mechanism, as I don't recall that off the top of my head).

I think my preference is that configuration of rez and rez plugins are maintained as closely as possible together, but that idea may get fuzzy if certain plugins end up being distributed to other sub-projects, like if rez-pip got split out, and required some configuration details.

**Reply by JeanChristopheMorinPerso** [reply]
> but that idea may get fuzzy if certain plugins end up being distributed to other sub-projects, like if rez-pip got split out, and required some configuration details.

That's a shortcoming of the current way the plugins system works. If plugins had a way to advertize their configuration schema to rez, then we could have the plugin settings declared in the main rez config and in the user `rezconfig.py`.

**Reply by bpabel** [reply]
@maxnbk For the merging, I was talking about how you can have multiple `rezconfig.py` files in the `REZ_CONFIG_FILE` path and their settings will get merged together.

Some of the plugin settings are already in the rezconfig, but only the ones that are universal across all plugins of that particular plugin type. It seemed confusing for plugins to have some of their configuration in rezconfig and some not in the rezconfig.
